servant-server/src/Servant/Server/Internal/Config.hs

```diff
@@ -0,0 +1,59 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE DeriveFoldable             #-}
+{-# LANGUAGE DeriveFunctor              #-}
+{-# LANGUAGE DeriveGeneric              #-}
+{-# LANGUAGE DeriveTraversable          #-}
+{-# LANGUAGE FlexibleInstances          #-}
+{-# LANGUAGE FunctionalDependencies     #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures             #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE PolyKinds                  #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TypeOperators              #-}
+{-# LANGUAGE UndecidableInstances       #-}
+
+#include "overlapping-compat.h"
+
+module Servant.Server.Internal.Config where
+
+import           Data.Proxy
+import           GHC.TypeLits
+
+-- | The entire configuration.
+data Config a where
+    EmptyConfig :: Config '[]
+    (:.) :: x -> Config xs -> Config (x ': xs)
+infixr 5 :.
+
+instance Show (Config '[]) where
+  show EmptyConfig = "EmptyConfig"
+instance (Show a, Show (Config as)) => Show (Config (a ': as)) where
+  showsPrec outerPrecedence (a :. as) =
+    showParen (outerPrecedence > 5) $
+      shows a . showString " :. " . shows as
```

Comment:
And less importantly, `Ord`. 

There is, I think, also only one sensible `Monoid` instance (delegating to all inner monoids), and that may serve some purpose to someone, but may not :). There is also a `Bounded` instance that makes sense, but I can't imagine a use-case for it.